size_t/ptrdiff_t transition part 1: test(1), .sh.match, init, macro, stk(3)

JohnoKing opened the pull request:
This is the first of a thirteen(!!) part patch series that enables ksh93 to operate within a 64-bit address space (currently dubbed thickfold). These changes were accomplished by fixing most (but not all) of the warnings that materialize during compilation when the following clang compiler flags are passed:
`-Wsign-compare -Wshorten-64-to-32 -Wsign-conversion -Wimplicit-int-conversion`

Originally, this patch was going to take the suggested approach of replacing int with `ssize_t`. While that does work in practice, it's a bad idea because POSIX does not guarantee `ssize_t` will accept any negative value besides `-1`, despite its signed nature[^1]:
> The type `ssize_t` shall be capable of storing values at least
> in the range [-1, {SSIZE_MAX}].

As such, for correctness and portability these patches will prefer usage of the C89 `ptrdiff_t`, which is practically guaranteed to accept the full range of negative numbers an int can accept. In practice, `ptrdiff_t` and `size_t` will be of the same size, being both 32-bit or both 64-bit. There are edge cases where this might not be so[^2], but I've chosen to use `ptrdiff_t` despite that since the caveats of such edge cases aren't nearly as bad as that of a platform's `ssize_t` potentially rejecting values lower than -1.
[^1]: https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/sys_types.h.html#tag_14_70
[^2]: I'll note there are also edge cases where `ssize_t` and `size_t` are of differing bit sizes, `size_t` being unsigned 64-bit and `ssize_t` being signed 32-bit. I presume this situation is extremely uncommon, but it's yet another reason why `ptrdiff_t` is a better choice.

In some areas `ssize_t` results from e.g. `sfvalue()` may end up being used with `ptrdiff_t` variables. This is not pedantically correct, but it's certainly better than the prior `int` hell status quo, wherein `ssize_t` was usually shortened to int.

Conspicuous changes with noteworthiness:
- Fixed many, many compiler warnings by adding a considerable number of casts and changing the types of more than quite a few variables. This is a consequence of using compiler warnings as an aid for implementing 64-bit memory allocation; while the warnings were useful, there were so many warnings fixed in the process that it increased the patch sizes greatly.
- Transitioned the `strgrpmatch()` and `strngrpmatch()` calls to use a `ssize_t*` pointer rather than an `int*` pointer.
- For the `sh_options` macros/function, enforce `uint64_t` as the main argument type to fix compiler warnings.
- The libast stk sublibrary already uses the `ssize_t/size_t` types, which made it rather easy to integrate proper `ptrdiff_t` usage. In fact, `stktell()` has **always** returned a `ptrdiff_t` result. The documentation in stk(3) has been incorrectly claiming since < 1995 it returns `int`, which is both pedantically and in actuality wrong. That error has been rectified alongside the other updates to stk.

The thickfold patch series was accomplished by fixing compiler warnings, so I've decided to provide metrics as to the change in the total number of warnings occurring during compilation.
Change in the number of warnings on Linux when compiling with clang using `-Wsign-compare -Wshorten-64-to-32 -Wsign-conversion -Wimplicit-int-conversion` (using the 'warnings generated' strings):
4,085 => 3,852 => 60 (progression from 54ff4ec06 => part 1 => part 13)

Side note: To re-emphasize, this patch is one part of a whole (although it can be used on its own). The full suite of changes can be found on the thickfold-size_t branch. This first part has been submitted severed from the other changes to make code review less ~hellish~ laborious (I hope). The full patch series can be found on my branch (last updated 2026-03-04):
https://github.com/JohnoKing/ksh/tree/0cbfe823/patches

Progresses https://github.com/ksh93/ksh/issues/592

JohnoKing committed:
size_t/ptrdiff_t transition part 1: test(1), .sh.match, macro expansion, init and stk(3)

This is the first of a thirteen(!!) part patch series that enables
ksh93 to operate within a 64-bit address space (currently dubbed
thickfold). These changes were accomplished by fixing most (but not
all) of the warnings that materialize during compilation when the
following clang compiler flags are passed:
'-Wsign-compare -Wshorten-64-to-32 -Wsign-conversion -Wimplicit-int-conversion'

Originally, this patch was going to take the suggested approach of
replacing int with ssize_t. While that does work in practice, it's
a bad idea because POSIX does not guarantee ssize_t will accept
any negative value besides -1, despite its signed nature[1]:
    > The type ssize_t shall be capable of storing values at least
    > in the range [-1, {SSIZE_MAX}].
As such, for correctness and portability these patches will prefer
usage of the C89 ptrdiff_t, which is practically guaranteed to
accept the full range of negative numbers an int can accept. In
practice, ptrdiff_t and size_t will be of the same size, being
both 32-bit or both 64-bit. There are edge cases where this might
not be so, but I've chosen to use ptrdiff_t despite that since
the caveats of such edge cases aren't nearly as bad as that of
a platform's ssize_t rejecting values lower than -1.
[1]: https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/sys_types.h.html#tag_14_70

In some areas ssize_t results from e.g. sfvalue() may end up being
used with ptrdiff_t variables. This is not pedantically correct,
but it's certainly better than the prior int hell status quo,
wherein ssize_t was usually shortened to int.

Conspicuous changes with noteworthyness:
- Fixed many, many compiler warnings by adding a considerable
  number of casts and changing the types of more than quite
  a few variables. This is a consequence of using compiler
  warnings as an aid for implementing 64-bit memory allocation;
  while the warnings were useful, there were so many warnings
  fixed in the process that it increased the patch sizes greatly.
- Transitioned the strgrpmatch() and strngrpmatch calls to use
  a ssize_t* pointer rather than an int* pointer.
- For the sh_options macros/function, enforce uint64_t as the main
  argument type to fix compiler warnings.
- The libast stk sublibrary already uses the ssize_t/size_t types,
  which made it rather easy to integrate proper ptrdiff_t usage.
  In fact, stktell() has *always* returned a 'ptrdiff_t' result.
  The documentation in stk(3) has been incorrectly claiming
  since < 1995 it returns 'int', which is wrong. That error
  has been rectified alongside the other updates to stk.

The thickfold patch series was accomplished by fixing compiler warnings,
so I've decided to provide metrics as to the change in the total number
of warnings occurring during compilation.
Change in the number of warnings on Linux when compiling with clang using
-Wsign-compare -Wshorten-64-to-32 -Wsign-conversion -Wimplicit-int-conversion:
4,042 => 3,808 => 248 (progression from cc5e0692 => part 1 => part 13)

Side note: To re-emphasize, this patch is one part of a whole
(although it can be used on it's own). The full suite of
changes can be found on the thickfold-size_t branch.
This first part has been submitted severed from the other
changes to make code review less laborious (I hope).
The full patch series can be found here:
https://github.com/JohnoKing/ksh/tree/ad588c003d/patches

Progresses https://github.com/ksh93/ksh/issues/592

JohnoKing committed:
Trivial Cygwin warning fix

JohnoKing committed:
preempt possible merge conflict

This code is set to be deleted to fix ACE vulns,
so undo the ptrdiff_t change to avoid a merge conflict.

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into thickfold-part-1-submission

JohnoKing committed:
Update dates

McDutchie commented:
> As such, for correctness and portability these patches will prefer usage of the C89 `ptrdiff_t`, which is practically guaranteed to accept the full range of negative numbers an int can accept.

However, ptrdiff_t is meant to store the difference between two pointers (hence the name) so it's quite misleading, in terms of reading the code, to use it as a general int type.

There is intmax_t and uintmax_t, provided by libast on old systems that don't have these natively. Why not use those?

McDutchie commented:
Hmmm… did you choose `ptrdiff_t` because it is a 64-bit int type on 64-bit systems and a 32-bit int type on 32-bit systems?

McDutchie commented:
OK, that seems technically sound in practice, but `ptrdiff_t` is specifically intended for storing the result of subtracting two pointers, which is just confusing when reading the code if it's used for all sorts of integer values.

As far as I can tell, using `intptr_t` (or `uintptr_t` for unsigned) would work as well, and it seems a more explicit/elegant way of using an integer type that is the size of a pointer. Their availability is guaranteed by libast in features/common.

JohnoKing commented:
It's also worth noting on ARM CHERI `intptr_t` is 128-bit (and contains more than the pointer address) whilst `ptrdiff_t` is 64-bit. In that case you still only have 64-bits for addressable memory, meaning `intptr_t` and `ptrdiff_t` aren't interchangeable.

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into thickfold-part-1-submission

JohnoKing committed:
Sync with thickfold-size_t branch

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into thickfold-part-1-submission

JohnoKing committed:
Fix a -Wincompatible-pointer-types-discards-qualifiers warning

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into thickfold-part-1-submission

JohnoKing committed:
Improve size_t casts to prefer converting additive results

JohnoKing committed:
Merge remote-tracking branch 'upstream/dev' into thickfold-part-1-submission

JohnoKing committed:
Update to use signed_size_t

JohnoKing commented:
I managed to refactor the PR to use a `signed_size_t` type (see commit message for explanation/rationale): https://github.com/JohnoKing/ksh/commit/cde9b44206e8de235db2ae271d132a6fb46c3052

McDutchie committed:
Merge branch 'dev' into thickfold-part-1-submission

McDutchie commented:
As far as I can tell, in the current iteration of the PR, `signed_size_t` is exactly the same as the POSIX `ssize_t` type; your feature test ensures that `signed_size_t` is the same as `size_t`, but that also applies to `ssize_t`. Is there any reason not to use `ssize_t` instead?

JohnoKing commented:
> `signed_size_t` is exactly the same as the POSIX `ssize_t`

Not in all cases. The bitness of `ssize_t` and `size_t` is not guaranteed to match (it's possible to have a 32-bit `ssize_t` and 64-bit `size_t`). Additionally, the POSIX standard doesn't guarantee `ssize_t` will accept anything lower than -1:

> The type ssize_t shall be capable of storing values at least
> in the range [-1, {SSIZE_MAX}].

This is why I'm using `signed_size_t` instead for the PR (at least unless some future version of POSIX guarantees the bitness of `size_t` and `ssize_t` match + adds a `SSIZE_MIN` definition; I'm not counting on it).

McDutchie commented:
(sigh) Right, of course, I'd forgotten about that insane POSIX definition of `ssize_t`.

It's extremely annoying though because I'm 99.9% sure there isn't an extant system where `ssize_t` is not simply the signed version of `size_t`.

Also note that ksh93 has always assumed a 32-bit `int` type, and that is not guaranteed by either the C standard or POSIX either (`int` is permitted to be 16 bits wide).

Ah well, we can always change it again later. Maybe I'll raise it at the Austin Group.

McDutchie committed:
Merge branch 'dev' into thickfold-part-1-submission

McDutchie commented on src/cmd/ksh93/bltins/test.c:
```diff
@@ -165,7 +166,7 @@ int b_test(int argc, char *argv[],Shbltin_t *context)
 		case 4:
 		{
-			int op = sh_lookup(cp=argv[2],shtab_testops);
+			uint64_t op = sh_lookup(cp=argv[2],shtab_testops);
```
Since the type returned by `sh_lookup` is `const unsigned` a.k.a. `const unsigned int` (see shtable.h), I don't understand why a 64-bit unsigned int is needed here. To fix any compiler warnings, `unsigned int` should be enough, no? (Same with `bop` below)

McDutchie committed:
features/common: revert size_t changes, fail more gracefully

- Using size_t here makes little sense; the max value is 8! The
  original code was fine.
- Use iffe's NOTE() macro and return 1 to error out more
  gracefully. (NOTE() uses write(2), which requires unistd.h.)
- Update the test's title and failure message.

JohnoKing committed:
Fix warnings with more minimal changes (re: 7c8fc3e)

The type was not changed to size_t because of the actual size, but
rather to fix -Wsign-conversion warnings. Since signed int variables are
apparently more desirable, (int) casts can be used instead.
(This change is not essential; it's merely been committed to fix
unimportant compiler warnings.)

JohnoKing committed:
Reduce op/bop to 32-bit

https://github.com/ksh93/ksh/pull/867#discussion_r2856597119

McDutchie committed:
Generate signed_size_t more elegantly

We already have the _ast_intSIZE_t macros for signed integer types
of a specified size, so we can just use one of those, with SIZE
being sizeof(size_t).

McDutchie commented:
I'm sorry to be a pain here, but there's still a couple of things I would like to be clearer about before committing this.

First is, you've added a lot of explicit typecasts from one integer type to another. My question is: are these necessary to avoid compiler warnings, or are they just a stylistic choice? My understanding was always that it is okay to leave typecasts between integer types implicit.

I'm not really a fan of the explicit typecasts, because they make the code more verbose, and potentially add to the maintenance load if the argument types to a function need to be changed.

Second, I've been looking into the C standard's `size_t` specification and POSIX's `ssize_t` specification. I can't find anything in the standards that guarantees that these types are more than 16 bits wide, even on 64-bit systems. As far as I can tell, it would be perfectly compliant to have a 64-bit system with 16 exabytes of RAM where the maximum object size returned by `sizeof` is 65535 bytes and `size_t` is 16 bits wide.

So I'm questioning whether the use of `signed_size_t` really makes any meaningful progress towards standards compliance. The way the standards specify these types, it looks to me like you cannot even rely on `size_t` being useful in the first place on modern systems.

So I'm coming to the impression that, since these types appear to be underspecified, pedantic standards compliance here is irrelevant (because it looks impossible anyway) and that what matters is real-world portability.

In the real world, AFAIK, there is no known system where `ssize_t` cannot contain values \< -1, because all signed integer types use the most significant bit for the sign. And in fact the AST/ksh code base was happily assuming the ability to pass values \< -1 in (at least) `stkseek` and it was working just fine on everything (though I agree that using `ptrdiff_t` is more correct here).

Nor, AFAIK, is there any system where `size_t` and `ssize_t` are different widths.

So is there any real reason why we shouldn't simply use `ssize_t` instead of `signed_size_t` in this PR? I'm not seeing one, but of course I could be wrong.

JohnoKing committed:
Use ssize_t rather than pedantic signed_size_t

This is done by checking out the thickfold-size_t branch's files.

JohnoKing commented:
> you've added a lot of explicit typecasts from one integer type to another. My question is: are these necessary to avoid compiler warnings, or are they just a stylistic choice

The casts were necessary to silence a huge number of them. Throughout the codebase ksh was frequently doing implicit int shortening, which caused memory allocation to fail with 64-bit sizes, so to fix those reliably I had to eliminate thousands of superfluous warnings (cruft) to identify *where* the memory allocation bugs were actually located in the codebase.

The codebases's tendency to implicitly shorten integers is also a source of bugs/problems that extend beyond just memory allocation. An example of this is https://github.com/att/ast/issues/1038, which is caused by ksh implicitly shortening 32-bit integers to 16-bit short integers. I fixed that bug in a [separate patch](https://github.com/JohnoKing/ksh/commit/dda6c9a5b718d02f791fa56a26497f698b0b3327), though that hasn't been submitted because the nvflags patch requires changes implemented in the thickfold patches (so this PR + the [12 other monster patches](https://github.com/JohnoKing/ksh/tree/0cbfe823/patches)).

> So I'm coming to the impression that, since these types appear to be underspecified, pedantic standards compliance here is irrelevant (because it looks impossible anyway) and that what matters is real-world portability.

I concede this is probably a better approach. I've dropped `signed_size_t` in favor of `ssize_t`. Anything with a weird `ssize_t` probably didn't work well to begin with.

McDutchie committed:
Merge branch 'dev' into thickfold-part-1-submission

McDutchie commented:
> In the real world, AFAIK, there is no known system where `ssize_t` cannot contain values < -1, because all signed integer types use the most significant bit for the sign. And in fact the AST/ksh code base was happily assuming the ability to pass values < -1 in (at least) `stkseek` and it was working just fine on everything (though I agree that using `ptrdiff_t` is more correct here).

I was about to commit this and then I noticed that my comment, above, about `stkseek` is completely wrong. Its `ssize_t` argument must be an absolute offset relative to `stream->_data` (see stk.c:378) so it's not clear to me why the type is even `ssize_t` and not `size_t` (in fact it was changed to that from `unsigned` (int) in 2011). Anyway that is not evidence that the code base was assuming the ability of `ssize_t` to have values \< -1. I will look for other such evidence but if I can't find any, I think we should probably reintroduce `signed_size_t` after all.

JohnoKing commented:
>  I will look for other such evidence but if I can't find any, I think we should probably reintroduce `signed_size_t` after all.

I found this in libast iconv:
https://github.com/ksh93/ksh/blob/a215b8be7958974902299233231611165b132114/src/lib/libast/comp/iconv.c#L1214-L1215


Edit: Another find in dthash:
https://github.com/ksh93/ksh/blob/a215b8be7958974902299233231611165b132114/src/lib/libast/cdt/dthash.c#L45-L61

McDutchie commented:
Thanks. That'll do for `ssize_t`, especially the second example.

McDutchie commented:
Found another problem: the new `sh_mactrim` declaration assumes that `char` is signed and can contain a negative value, but that's unspecified: `char` may be either signed or unsigned by default, depending on the implementation. The `mode` argument should either revert to `int` or be explicitly specified as `signed char`. I think the former is probably better style-wise. I'll push a commit.

McDutchie committed:
fix assumption of signed char in sh_mactrim

The new sh_mactrim declaration assumes that char is signed and can
contain a negative value, but that's unspecified: char may be
either signed or unsigned by default, depending on the
implementation. The mode argument should either revert to int or be
explicitly specified as signed char. I think the former is probably
better style-wise.

Changed the somewhat strange 'mode * -1' to equivalent '-mode'.
Added an explicit typecast to avoid a potential compiler warning.
